### PR TITLE
Split into vorticity and buoyancy diffusion

### PR DESCRIPTION
**Pull request by matt-frey**
We generalise the diffusion such that we can have different diffusion settings for vorticity and buoyancy. We also add some more diagnostics to the netcdf file.

### CODE REVIEW

File: src/stepper/advance.f90 [hunk 1]
```diff
@@ -216,6 +229,13 @@ subroutine adapt(bstep, t, dt)
             call diffy(svel(:, :, :, 3), ys)
             call fftxys2p(ys, dwdy)
 
+
+            ! Upper surface strain rms:
+            ! S = 0.5 * sqrt{(u_y + v_x)^2 + (u_x - v_y)^2}
+            ! v_x = u_y + zeta
+            ussrms = sum((two * dudy(nz, :, :) +  vor(nz, :, :, 3)) ** 2 + &
```
Comment (daviddritschel) [inline src/stepper/advance.f90:236]: I think you are missing the 0.5 multiplying sum; look at the comment 2 lines earlier

File: src/stepper/cn2.f90 [hunk 2]
```diff
@@ -55,8 +61,24 @@ subroutine cn2_set_diffusion(self, dt, vorch)
              endif
 
             !$omp parallel workshare
-            diss = one / (one + dfac * hdis)
+            vdiss = one / (one + dfac * vhdis)
```
Comment (daviddritschel) [inline src/stepper/cn2.f90:64]: This does not look like the usual exponential integrating factor (see my code, evolution.f90).  Is this ok?

File: src/stepper/cn2.f90 [hunk 3]
```diff
@@ -55,8 +61,24 @@ subroutine cn2_set_diffusion(self, dt, vorch)
...
+                !Update diffusion operator used in time stepping:
+                dbac = dt
+            else
+                !Update hyperdiffusion operator used in time stepping:
+                dbac = bf * dt
+             endif
+
+
+            !$omp parallel workshare
+            bdiss = one / (one + dbac * bhdis)
```
Comment (daviddritschel) [inline src/stepper/cn2.f90:79]: Same here as previous comment

File: src/stepper/impl_rk4.f90 [hunk 4]
```diff
@@ -77,23 +84,30 @@ subroutine impl_rk4_step(self, t, dt)
             dt6 = f16 * dt
 
             ! set integrating factors
-            self%emq = dexp(- diss)
+            self%emq = dexp(- vdiss)
```
Comment (daviddritschel) [inline src/stepper/impl_rk4.f90:87]: ok, this looks like the integrating factor.  The other comments may pertain to vertical filtering?

**Review by daviddritschel — COMMENTED** [review]
Just a few comments, one significant (factor of 2)

File: src/stepper/advance.f90 (hunk 1 above)
Comment (matt-frey) [reply src/stepper/advance.f90:236]: I multiply later by 1/2 after getting the global sum. See https://github.com/matt-frey/ps3d/pull/71/files#diff-cef68695f14ebfde89f7eb2a68a4ea7636ec47f02124fe6c327b674589d1a0d7R318

File: src/stepper/cn2.f90 (hunk 2 above)
Comment (matt-frey) [reply src/stepper/cn2.f90:64]: That is another integrator. We provide two time integrators.

File: src/stepper/cn2.f90 (hunk 3 above)
Comment (matt-frey) [reply src/stepper/cn2.f90:79]: I can close this as this is not the time integrator using integrating factors.

File: src/stepper/impl_rk4.f90 (hunk 4 above)
Comment (matt-frey) [reply src/stepper/impl_rk4.f90:87]: We have cn2.f90 (the integrator we used for the Beltrami paper) and impl_rk4.f90 (the integrator using integrating factors for diffusion).

**Review by daviddritschel — COMMENTED** [review]
OK, I see that now.

File: src/stepper/cn2.f90 (hunk 3 above)
Comment (daviddritschel) [reply src/stepper/cn2.f90:79]: Yes, I think that is a good idea - it is only the leading order approximation to the integrating factor

**Pull request merged**